components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,352 @@
+        thread::{self, sleep},
+        time::Duration,
+    };
 
     use super::*;
+    use crate::time::Instant;
 
-    static COUNTER: AtomicUsize = AtomicUsize::new(0);
+    fn replace_pool_rule(thread_count: usize, thread_name: &str) -> TokioResult<Runtime> {
+        let rt = tokio::runtime::Builder::new_multi_thread()
+            .worker_threads(thread_count)
+            .thread_name(thread_name)
+            .enable_all()
+            .build()
+            .unwrap();
+        Ok(rt)
+    }
 
     #[test]
     fn test_adjust_thread_num() {
-        let replace_pool_rule = |thread_count: usize, thread_name: &str| {
-            let rt = tokio::runtime::Builder::new_multi_thread()
-                .worker_threads(thread_count)
-                .thread_name(thread_name)
-                .enable_all()
-                .build()
-                .unwrap();
-            Ok(rt)
+        static COUNTER: AtomicUsize = AtomicUsize::new(4);
+        let after_adjust = |new_size: usize| {
+            COUNTER.store(new_size, Ordering::SeqCst);
         };
+        let mut threads = ResizableRuntime::new(
+            COUNTER.load(Ordering::SeqCst),
+            "test",
+            Box::new(replace_pool_rule),
+            Box::new(after_adjust),
+        );
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size());
+
+        let handle = threads.handle();
+        handle.block_on(async {
+            COUNTER.fetch_add(1, Ordering::SeqCst);
+        });
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size() + 1);
+
+        threads.adjust_with(8);
+        assert!(!threads.gc_runtime.tracker.is_empty());
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size());
+
+        sleep(Duration::from_secs(1));
+        assert!(threads.gc_runtime.tracker.is_empty());
+
+        // New task should be scheduled to the new runtime
+        handle.block_on(async {
+            COUNTER.fetch_add(1, Ordering::SeqCst);
+        });
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size() + 1);
+    }
+
+    #[test]
+    fn test_infinite_loop() {
+        static COUNTER: AtomicUsize = AtomicUsize::new(4);
         let after_adjust = |new_size: usize| {
             COUNTER.store(new_size, Ordering::SeqCst);
         };
-        let mut threads =
-            ResizableRuntime::new("test", Box::new(replace_pool_rule), Box::new(after_adjust));
-        threads.adjust_with(4);
-        assert_eq!(COUNTER.load(Ordering::SeqCst), 4);
+        let mut threads = ResizableRuntime::new(
+            COUNTER.load(Ordering::SeqCst),
+            "test",
+            Box::new(replace_pool_rule),
+            Box::new(after_adjust),
+        );
+        assert_eq!(COUNTER.load(Ordering::SeqCst), threads.size());
+
+        let handle = threads.handle();
+        // infinite loop should not be cleaned
+        handle.spawn(async {
+            loop {
+                sleep(Duration::from_secs(10));
+            }
+        });
```

Comment:
you can actually just spawn `std::future::pending()` to create a task that never completes.

(ditto for all those `LONG_ENOUGH_TIME` below)

https://doc.rust-lang.org/std/future/fn.pending.html

Reply:
done